openedx/core/djangoapps/session_inactivity_timeout/middleware.py

```diff
@@ -34,27 +38,66 @@ def process_request(self, request):
+            # .. setting_name: SESSION_SAVE_FREQUENCY_SECONDS
+            # .. setting_default: 900
+            # .. setting_description: How often to allow a full session save (in seconds).
+            #   This controls how frequently the session ID might change
+            #   A user could be inactive for almost SESSION_SAVE_FREQUENCY_SECONDS but since their session 
+            #   isn't being saved during that time, their last activity timestamp isn't being updated.
+            #   When they hit the inactivity timeout, it will be based on the last saved activity time
+            #   So the effective timeout could be as short as: SESSION_INACTIVITY_TIMEOUT_IN_SECONDS - SESSION_SAVE_FREQUENCY_SECONDS
+            #   This means users might be logged out earlier than expected in some edge cases.
+            # .. setting_warning:  Used in sync with SESSION_INACTIVITY_TIMEOUT_IN_SECONDS
+            frequency_time_in_seconds = getattr(settings, "SESSION_SAVE_FREQUENCY_SECONDS", 900)
             # what time is it now?
             utc_now = datetime.utcnow()
 
             # Get the last time user made a request to server, which is stored in session data
-            last_touch = request.session.get(LAST_TOUCH_KEYNAME)
+            last_touch_str = request.session.get(LAST_TOUCH_KEYNAME)
 
             # have we stored a 'last visited' in session? NOTE: first time access after login
             # this key will not be present in the session data
-            if last_touch:
-                # compute the delta since last time user came to the server
-                time_since_last_activity = utc_now - last_touch
-
-                # did we exceed the timeout limit?
-                if time_since_last_activity > timedelta(seconds=timeout_in_seconds):
-                    # yes? Then log the user out
-                    del request.session[LAST_TOUCH_KEYNAME]
-                    auth.logout(request)
-                    return
-
-            request.session[LAST_TOUCH_KEYNAME] = utc_now
+            if last_touch_str:
+                try:
+                    # Convert the ISO string back to a datetime object
+                    last_touch = datetime.fromisoformat(last_touch_str)
+
+                    # compute the delta since last time user came to the server
+                    time_since_last_activity = utc_now - last_touch
+
+                    # did we exceed the timeout limit?
+                    if time_since_last_activity > timedelta(seconds=timeout_in_seconds):
+                        # yes? Then log the user out
+                        del request.session[LAST_TOUCH_KEYNAME]
+                        auth.logout(request)
+                        return
+                except (ValueError, TypeError) as e:
+                    # If parsing fails, treat as if no timestamp exists
+                    pass
+            else:
+                LOG.info("No previous activity timestamp found (first login)")
```

Comment:
1. I think this would cause too mamy log messages at info level.
```suggestion
                LOG.debug("No previous activity timestamp found (first login)")
```
2. You could use `set_custom_attribute` if you want to add additional observability. That's up to you. Some possible keys include:
  - `session_inactivity.activity_seen`
  - `session_inactivity.first_login`
  - `session_inactivity.session_extended`
See https://github.com/openedx/auth-backends/pull/382 for examples.

Reply:
updated